libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -330,6 +364,45 @@ proc canAskIWant*(p: PubSubPeer, msgId: MessageId): bool =
       return true
   return false
 
+proc clearSendPriorityQueue(p: PubSubPeer) =
+  while p.rpcmessagequeue.sendPriorityQueue.len > 0 and p.rpcmessagequeue.sendPriorityQueue[0].finished:
+    when defined(libp2p_expensive_metrics):
+      libp2p_gossipsub_priority_queue_size.dec(labelValues = [$p.peerId])
+    discard p.rpcmessagequeue.sendPriorityQueue.popFirst()
+
+proc sendNonPriorityTask(p: PubSubPeer) {.async.} =
+  while true:
+     while p.rpcmessagequeue.sendPriorityQueue.len > 0:
+       await p.rpcmessagequeue.sendPriorityQueue[0]
```

Comment:
it's better to wait for the last item - if you have 5 items here that are sent in order, you will necessarily have to wait for all of them.

also, the queue should be cleared before await so that you don't await a finished future unnecessarily

Reply:
Could please elaborate on what's the difference between waiting the first or the last? The current approach seems more intuitive to me. 

If we clear first we also might call the clear proc unnecessarily as the first future might not have finished yet.

Reply:
the logic should be:
* remove all finished futures from the queue
* if the queue is not empty, wait for the _last_ remaining future
* repeat until empty

this minimizes the number of times we have to wait for something (each wait = performance cost) - we will never wait for a finished future and by waiting for the last one, all that come before it are guaranteed to be finished already (since sends are processed in order).

Reply:
thanks for explaining it, it makes sense.